src/loss.cpp

```diff
@@ -0,0 +1,83 @@
+#include <vector>
+
+#include <Eigen/Core>
+#include <deepworks/loss.hpp>
+#include "util/assert.hpp"
+
+using ConstMatrix = Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+using Matrix = Eigen::Map<Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+
+using ConstVector = Eigen::Map<const Eigen::Matrix<float, 1, Eigen::Dynamic, Eigen::RowMajor>>;
+using Vector = Eigen::Map<Eigen::Matrix<float, 1, Eigen::Dynamic, Eigen::RowMajor>>;
+
+
+namespace deepworks {
+namespace loss {
+
+std::vector<int> GetTargetIndices(const ConstVector target, int batch_size, int n_classes) {
+    std::vector<int> slice(batch_size);
+    for (int i = 0; i < batch_size; ++i) {
+        slice[i] = static_cast<int>(target(0, i)) + n_classes * i;
+    }
+    return slice;
+}
+
+float CPUNLLLoss(Matrix X, const ConstVector target) {
+    int batch_size = X.rows();
+    int n_classes = X.cols();
+
+    Vector X_1d(X.data(), X.size());
+
+    std::vector<int> slice = GetTargetIndices(target, batch_size, n_classes);
+
+    float loss = -X_1d(0, slice).array().sum() / static_cast<float>(batch_size);
+    return loss;
+}
+
+void CPULog(const ConstMatrix X, Matrix LogX) {
+    LogX.array() = X.array().log();
+}
+
+float CPUCrossEntropyLossForward(const Tensor& X, const Tensor& target) {
+    const auto &shape = X.shape();
+    DeepWorks_Assert(shape.size() == 2);
+
+    int batch_size = shape[0];
+    int n_classes = shape[1];
+
+    ConstMatrix prob_mat(X.data(), batch_size, n_classes);
+    ConstVector target_vec(target.data(), batch_size);
+
+    Eigen::MatrixXf log_prob(batch_size, n_classes);
+    Matrix log_prob_mat(log_prob.data(), batch_size, n_classes);
```

Comment:
Why do we need `Eigen::MatrixXf log_prob`?

Reply:
> Why do we need `Eigen::MatrixXf log_prob`?

Our `Matrix` is the `Eigen::Map`, in fact, it's only `view` on data. We need to allocate memory for `Log Matrix`, cause X is constant, then i use MatrixXf.